baseband/mark5b/header.py

```diff
@@ -254,13 +290,54 @@ def get_time(self, framerate=None, frame_nr=None):
         return Time(self.kday + self.jday, (self.seconds + offset) / 86400,
                     format='mjd', scale='utc', precision=9)
 
-    def set_time(self, time):
+    def set_time(self, time, framerate=None, frame_nr=None):
+        """
+        Convert Time object to BCD timestamp elements and frame_nr.
+
+        For non-integer seconds, the frame_nr will be calculated if not given
+        explicitly. Doing so requires the frame rate.
+
+        Parameters
+        ----------
+        time : Time instance
+            The time to use for this header.
+        framerate : `~astropy.units.Quantity`, optional
+            For calculating the ``frame_nr`` from the fractional seconds.
+        frame_nr : int, optional
+            An explicit frame number associated with the fractions of seconds.
+        """
         self.kday = int(time.mjd // 1000) * 1000
         self.jday = int(time.mjd - self.kday)
-        ns = int(round((time - Time(self.kday + self.jday, format='mjd')).sec *
-                       1e9))
-        sec, ns = divmod(ns, 1000000000)
-        self.seconds = sec
-        self.ns = ns
+        seconds = time - Time(self.kday + self.jday, format='mjd')
+        int_sec = int(seconds.sec)
+        frac_sec = seconds - int_sec * u.s
+
+        # Round to nearest 2 ns to handle timestamp difference errors.
+        if abs(frac_sec) < 2. * u.ns:
+            frame_nr = 0
+            ns = 0.
+        elif abs(1. * u.s - frac_sec) < 2. * u.ns:
```

Comment:
Let's at least keep the rounding consistent!

Reply:
I'm not actually sure what's inconsistent with it.  Could you clarify?

Reply:
Just that you use `1 * u.ns` further down.